src/test/test_perf_counters_cache.cc

```diff
@@ -0,0 +1,816 @@
+#include "common/perf_counters_cache.h"
+#include "common/perf_counters_key.h"
+#include "common/admin_socket_client.h"
+#include "global/global_context.h"
+#include "global/global_init.h"
+#include "include/msgr.h" // for CEPH_ENTITY_TYPE_CLIENT
+#include "gtest/gtest.h"
+
+using namespace std;
+
+int main(int argc, char **argv) {
+  std::map<string,string> defaults = {
+    { "admin_socket", get_rand_socket_path() }
+  };
+  std::vector<const char*> args;
+  auto cct = global_init(&defaults, args, CEPH_ENTITY_TYPE_CLIENT,
+			 CODE_ENVIRONMENT_UTILITY,
+			 CINIT_FLAG_NO_DEFAULT_CONFIG_FILE|
+			 CINIT_FLAG_NO_CCT_PERF_COUNTERS);
+  common_init_finish(g_ceph_context);
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
+
+enum {
+  TEST_PERFCOUNTERS1_ELEMENT_FIRST = 200,
+  TEST_PERFCOUNTERS_COUNTER,
+  TEST_PERFCOUNTERS_TIME,
+  TEST_PERFCOUNTERS_TIME_AVG,
+  TEST_PERFCOUNTERS1_ELEMENT_LAST,
+};
+
+std::string sd(const char *c)
+{
+  std::string ret(c);
+  std::string::size_type sz = ret.size();
+  for (std::string::size_type i = 0; i < sz; ++i) {
+    if (ret[i] == '\'') {
+      ret[i] = '\"';
+    }
+  }
+  return ret;
+}
+
+void add_test_counters(PerfCountersBuilder *pcb) {
+  pcb->add_u64(TEST_PERFCOUNTERS_COUNTER, "test_counter");
+  pcb->add_time(TEST_PERFCOUNTERS_TIME, "test_time");
+  pcb->add_time_avg(TEST_PERFCOUNTERS_TIME_AVG, "test_time_avg");
+}
+
+
+static PerfCountersCache* setup_test_perf_counters_cache(CephContext *cct, uint64_t target_size = 100)
+{
+  std::function<void(PerfCountersBuilder*)> lpcb = add_test_counters;
+  CountersSetup test_counters_setup(TEST_PERFCOUNTERS1_ELEMENT_FIRST, TEST_PERFCOUNTERS1_ELEMENT_LAST, lpcb);
+  std::unordered_map<std::string_view, CountersSetup> setups;
+  setups["key1"] = test_counters_setup;
+  setups["key2"] = test_counters_setup;
+  setups["key3"] = test_counters_setup;
+  setups["key4"] = test_counters_setup;
+  setups["key5"] = test_counters_setup;
+  setups["key6"] = test_counters_setup;
+  setups["good_ctrs"] = test_counters_setup;
+  setups["bad_ctrs1"] = test_counters_setup;
+  setups["bad_ctrs2"] = test_counters_setup;
+  setups["bad_ctrs3"] = test_counters_setup;
+  setups["too_many_delimiters"] = test_counters_setup;
+  return new PerfCountersCache(cct, target_size, setups);
+}
+
+void cleanup_test(PerfCountersCache *pcc) {
+  pcc->clear_cache();
+  delete pcc;
```

Comment:
is there any reason that the destructor can't clear the cache itself? cleaning up and releasing all of its memory is normally the job of the destructor